src/sentry/testutils/pytest/sentry.py

```diff
@@ -372,43 +371,23 @@ def pytest_collection_modifyitems(config, items):
 
     total_groups = int(os.environ.get("TOTAL_TEST_GROUPS", 1))
     current_group = int(os.environ.get("TEST_GROUP", 0))
-    grouping_strategy = os.environ.get("TEST_GROUP_STRATEGY", "file")
+    grouping_strategy = os.environ.get("TEST_GROUP_STRATEGY", "scope")
 
-    accepted, keep, discard = [], [], []
-
-    for index, item in enumerate(items):
-        # XXX: For some reason tests in `tests/acceptance` are not being
-        # marked as snuba, so deselect test cases not a subclass of SnubaTestCase
-        if os.environ.get("RUN_SNUBA_TESTS_ONLY"):
-            import inspect
-
-            from sentry.testutils.cases import SnubaTestCase
-
-            if inspect.isclass(item.cls) and not issubclass(item.cls, SnubaTestCase):
-                # No need to group if we are deselecting this
-                discard.append(item)
-                continue
-            accepted.append(item)
+    groups = collections.defaultdict(list)
+    for item in items:
+        if grouping_strategy == "scope":
+            key = item.nodeid.rsplit("::")[0]
         else:
-            accepted.append(item)
-
-        # In the case where we group by round robin (e.g. TEST_GROUP_STRATEGY is not `file`),
-        # we want to only include items in `accepted` list
-        item_to_group = (
-            int(md5(str(item.location[0]).encode("utf-8")).hexdigest(), 16)
-            if grouping_strategy == "file"
-            else len(accepted) - 1
-        )
+            key = item.nodeid
 
-        # Split tests in different groups
-        group_num = item_to_group % total_groups
+        groups[key].append(item)
 
-        if group_num == current_group:
-            keep.append(item)
-        else:
-            discard.append(item)
+    buckets: list[list[pytest.Item]] = [[] for _ in range(total_groups)]
+    for group in sorted(groups.values(), key=lambda group: -len(group)):
+        min(buckets, key=lambda bucket: len(bucket)).extend(group)
```

Comment:
pretty neat binpacking -- pick the longest groups first and insert them into the shortest bucket